### ISSUE TITLE
mismatch between elm-test and elm-test-rs report time

### ISSUE THREAD

**Issue by aurelienbottazini** [mpizenberg/elm-test-rs#125]
conversation started on this reddit thread: 
https://www.reddit.com/r/elm/comments/11nll8e/comment/jd89iay/?utm_source=reddit&utm_medium=web2x&context=3

I think elm-test-rs reports compile time + runtime
I think elm-test reports only runtime

Thus it made me think that elm-test-rs was slower

![](https://raw.githubusercontent.com/aurelienbottazini/test-elm-test-runners/main/doc/test-results.jpg)

Not sure if there's something to do about it.
I think both tools are awesome.

**Comment by mpizenberg** [comment]
Thank you for taking the time to report this! Indeed both projects do not measure time exactly the same and it's probably something that should be documented better.

Because measuring program time in Elm is not straightforward, in both cases the time measured is done in the JS supervision code. Also in both cases, the time measured is the time spent running tests, and does not take into account the preparation time (parsing the tests, solving dependencies, building the test file, etc.). But where the main difference lies is how time is being measured.

In elm-test supervision code, a [start time is recorded just before running the test suite](https://github.com/rtfeldman/node-test-runner/blob/448eafc69fbff5fbb7ee844f15da1f7565c9a568/lib/Supervisor.js#L28), and the [end time is recorded after all workers are done running the tests](https://github.com/rtfeldman/node-test-runner/blob/448eafc69fbff5fbb7ee844f15da1f7565c9a568/lib/Supervisor.js#L177). The reported time is thus the total time spent running the tests in a global fashion.

In elm-test-rs, due to how the test runner is designed, there is a lot more coordination and messaging happening between the supervisor and the workers (this is to make debug log capture easier). I did not want reported runtime to be affected by all this. In addition, I wanted to be able to report detailed timings for each test run. So instead I measure runtime of each test ([start](https://github.com/mpizenberg/elm-test-rs/blob/d9e0c921c986311f80fdfccc7d2cd9502b9a9520/templates/node_runner.js#L26), [end](https://github.com/mpizenberg/elm-test-rs/blob/d9e0c921c986311f80fdfccc7d2cd9502b9a9520/templates/node_runner.js#L37)). The reported duration at the end is simply the sum of all individual test runs. So it is quite different to the one reported in elm-test because it does not care about the number of workers used. The time reported is likely the time that a single worker would spend running all the tests. You can check that running `elm-test-rs --workers N` with N is 1, 2, 4 etc does not change much the reported duration.

In fact, if I run your tests with `--workers 4` on my machine you can see that the total time spent in the command is smaller than what elm-test-rs display in duration.

```
❯ time elm-test-rs --workers 4

Running 5000 tests. To reproduce these results later,
run elm-test-rs with --seed 606925938 and --fuzz 100


TEST RUN PASSED

Duration: 2568 ms
Passed:   5000
Failed:   0

________________________________________________________
Executed in    2.14 secs    fish           external
   usr time    3.65 secs    0.00 micros    3.65 secs
   sys time    0.73 secs  900.00 micros    0.73 secs
```

But because it's also interesting to know the actual global "human" duration spent running the tests, this information is displayed in verbose mode.

```
> time elm-test-rs -v --workers 4
...
TEST RUN PASSED

Duration: 2528 ms
Passed:   5000
Failed:   0
Running duration (since Node.js start): 1352 ms


________________________________________________________
Executed in    2.22 secs    fish           external
   usr time    3.92 secs    0.00 micros    3.92 secs
   sys time    0.84 secs  853.00 micros    0.83 secs
```

In the above example, it reports that 1352 ms was spent in since the Node start. This includes spawning the workers, distributing the test load, gathering the results. Add to this roughly 100ms to start node, and we get the following timings for elm-test-rs:

- Total time: 2.2 s
  - preparation work in Rust (parsing tests, dependency solving, preparing the test file, elm compilation, ...): 0.8 s
  - running tests: 1.4 s

And in elm-test, this looks more like this:

- Total time: 2.8 s
  - preparation work in JS (parsing tests, dependency solving, preparing the test file, elm compilation, ...): 1.6 s
  - running tests: 1.2 s

So as you can see, elm-test-rs gains some preparation time thanks to Rust speed, but loose some run time due to more message passing. Usually these even out and both elm-test and elm-test-rs have roughly the same actual total timings.

In conclusion, I guess I could change a bit the reporting wording?

**Comment by aurelienbottazini** [comment]
thanks that's very interesting

I think the reporting times as so slow that a change in the wording / output is not necessary.
I would not do it so I closing the issue with this comment.

Feel free to reopen if you want of course (I don't want add an issue to the project)
